File: opcua-webservice-spring/core/src/main/java/org/opcfoundation/webserver/service/transactions/object/UaReadMemberAttributeTransaction.java
package org.opcfoundation.webserver.service.transactions.object;

import org.eclipse.milo.opcua.stack.core.AttributeId;
import org.eclipse.milo.opcua.stack.core.StatusCodes;
import org.eclipse.milo.opcua.stack.core.UaRuntimeException;
import org.eclipse.milo.opcua.stack.core.types.builtin.*;
import org.eclipse.milo.opcua.stack.core.types.builtin.unsigned.UByte;
import org.eclipse.milo.opcua.stack.core.types.builtin.unsigned.UInteger;
import org.eclipse.milo.opcua.stack.core.types.structured.ReadValueId;
import org.opcfoundation.webserver.addressspace.models.UaObjectType;
import org.opcfoundation.webserver.addressspace.nodemanager.NodeManagerWebService;
import org.opcfoundation.webserver.addressspace.nodes.*;
import org.opcfoundation.webserver.service.transactions.base.UaReadTransaction;
import org.opcfoundation.webserver.types.UaChildId;
import org.opcfoundation.webapi.service.types.ReadContext;
import org.opcfoundation.webserver.types.UaChildIdentifier;
import org.opcfoundation.webserver.types.UaObjectId;
import org.opcfoundation.webserver.types.UaObjectIdentifier;
import org.opcfoundation.webserver.types.message.ReadMemberAttributeRequest;
import org.opcfoundation.webserver.types.message.ReadMemberAttributeResponse;

import java.util.*;
import java.util.concurrent.CompletableFuture;

public class UaReadMemberAttributeTransaction extends UaReadTransaction {
    private final UaObjectIdentifier objectId;
    private final UaChildIdentifier memberId;
    private final NodeManagerWebService nodeManager;

    public UaReadMemberAttributeTransaction(
            ReadContext context,
            UaObjectIdentifier objectId,
            UaChildIdentifier memberId,
            List<Integer> handleIds,
            NodeManagerWebService nodeManager)
    {
        super(context,handleIds);
        this.objectId = objectId;
        this.memberId = memberId;
        this.nodeManager = nodeManager;
    }

    public CompletableFuture<Void> execute()
    {
        try {
            UaObjectType objectType = nodeManager.findObjectType(objectId);
            if (null == objectType) throw new UaRuntimeException(StatusCodes.Bad_NodeIdUnknown);

            UaObject instanceDeclaration = nodeManager.findInstanceDeclaration(objectId);
            ReadMemberAttributeRequest request = new ReadMemberAttributeRequest(
                    new UaObjectId(
                            objectId.getId(),
                            instanceDeclaration),
                    new UaChildId(memberId.getPath(), memberId.getPathL2()));

            return objectType.onReadMemberAttributes(request).
                    thenAccept(this::setResults).
                    exceptionally(ex -> buildErrorResponse(ex.getCause()));

        } catch (Exception e) {
            return CompletableFuture.completedFuture(buildErrorResponse(e));
        }
    }

    private void setResults(ReadMemberAttributeResponse response)
    {
        List<ReadValueId> nodesToRead = getRequestedItems();

        for (ReadValueId item: nodesToRead)
        {
            Variant value = Variant.NULL_VALUE;
            StatusCode statusCode = StatusCode.GOOD;

            if (item.getAttributeId().intValue() == AttributeId.NodeClass.id()) {
                value = new Variant(response.getNodeClass().getValue());
            } else if (item.getAttributeId().intValue() == AttributeId.DisplayName.id()) {
                value = new Variant(response.getDisplayName());
            } else if (item.getAttributeId().intValue() == AttributeId.Description.id()) {
                if (!response.getDescription().isNull())
                {
                    value = new Variant(response.getDescription());
                } else {
                    statusCode = StatusCode.of(StatusCodes.Bad_NodeAttributesInvalid);
                }
            } else if (item.getAttributeId().intValue() == AttributeId.DataType.id()) {
                if (null != response.getDataTypeId()) {
                    value = new Variant(response.getDataTypeId());
                } else {
                    statusCode = StatusCode.of(StatusCodes.Bad_NodeAttributesInvalid);
                }
            } else if (item.getAttributeId().intValue() == AttributeId.ValueRank.id()) {
                if (null != response.getValueRank()) {
                    value = new Variant(response.getValueRank());
                } else {
                    statusCode = StatusCode.of(StatusCodes.Bad_NodeAttributesInvalid);
                }
            } else if (item.getAttributeId().intValue() == AttributeId.AccessLevel.id()) {
                if (null != response.getAccessLevel()) {
                    value = new Variant(UByte.valueOf(response.getAccessLevel()));
                } else {
                    statusCode = StatusCode.of(StatusCodes.Bad_NodeAttributesInvalid);
                }
            } else if (item.getAttributeId().intValue() == AttributeId.UserAccessLevel.id()) {
                if (null != response.getAccessLevel()) {
                    value = new Variant(UByte.valueOf(response.getAccessLevel()));
                } else {
                    statusCode = StatusCode.of(StatusCodes.Bad_NodeAttributesInvalid);
                }
            } else if (item.getAttributeId().intValue() == AttributeId.Historizing.id()) {
                if (null != response.getHistorizing()) {
                    value = new Variant(response.getHistorizing());
                } else {
                    statusCode = StatusCode.of(StatusCodes.Bad_NodeAttributesInvalid);
                }
            } else if (item.getAttributeId().intValue() == AttributeId.NodeId.id()) {
                value = new Variant(item.getNodeId());
            } else if (item.getAttributeId().intValue() == AttributeId.BrowseName.id()) {
                value = new Variant(new QualifiedName(0, memberId.getPath()));
            } else if (item.getAttributeId().intValue() == AttributeId.Executable.id()) {
                value = new Variant(Boolean.TRUE);
            } else if (item.getAttributeId().intValue() == AttributeId.UserExecutable.id()) {
                value = new Variant(Boolean.TRUE);
            } else if (item.getAttributeId().intValue() == AttributeId.WriteMask.id()) {
                value = new Variant(UInteger.valueOf(0));
            } else if (item.getAttributeId().intValue() == AttributeId.UserWriteMask.id()) {
                value = new Variant(UInteger.valueOf(0));
            }  else {
                statusCode = StatusCode.of(StatusCodes.Bad_AttributeIdInvalid);
            }

            results.add(new DataValue(
                    value,
                    statusCode,
                    null,
                    null));
        }
    }
}

File: opcua-webservice-node/src/server/service/transactions/reactiveobject/UaReadMemberAttributeTransaction.ts
import { Attributes, StatusCodes } from "opcua-webapi";
import { makeUaStatusCode, UaDataValue, UaError, UaReadValueId, UaVariant, UaVariantType } from "opcua-webapi-ts";
import { ReadContext } from "../../..";
import { NodeManagerReactiveObject } from "../../../addressspace/nodemanager/NodeManagerReactiveObject";
import { UaChildIdentifier, UaChildId, UaObjectId, UaObjectIdentifier } from "../../../types";
import { ReadMemberAttributeRequest, ReadMemberAttributeResponse } from "../../message";
import { UaReadTransaction } from "../base/UaReadTransaction";

export class UaReadMemberAttributeTransaction extends UaReadTransaction {
    private readonly objectId: UaObjectIdentifier;
    private readonly memberId: UaChildIdentifier;
    private readonly nodeManager: NodeManagerReactiveObject;

    constructor(
        context: ReadContext,
        objectId: UaObjectIdentifier,
        memberId: UaChildIdentifier,
        handleIds: Array<number>,
        nodeManager: NodeManagerReactiveObject,
    ) {
        super(context, handleIds);
        this.objectId = objectId;
        this.memberId = memberId;
        this.nodeManager = nodeManager;
    }

    override async execute(): Promise<void> {
        try {
            const objectType = this.nodeManager.findObjectType(this.objectId);
            if (objectType === null) {
                throw new UaError(makeUaStatusCode(StatusCodes.BadNodeIdUnknown));
            }

            const instanceDeclaration = this.nodeManager.findInstanceDeclaration(this.objectId);
            const request = new ReadMemberAttributeRequest(
                new UaObjectId(this.objectId.id, instanceDeclaration),
                new UaChildId(this.memberId.path, this.memberId.pathL2),
                this.memberId.methodNode,
            );

            const response = await objectType.onReadMemberAttributes(request);
            this.setResults(response);
        } catch (error) {
            this.buildErrorResults(error);
        }
    }

    private setResults(response: ReadMemberAttributeResponse): void {
        const nodesToRead = this.getRequestedItems();
        this._results = [];

        for (const item of nodesToRead) {
            let value = UaVariant.null();
            let statusCode = makeUaStatusCode(StatusCodes.Good);

            if (item.attributeId === Attributes.NodeClass) {
                value = UaVariant.integer(response.nodeClass, UaVariantType.UInt32);
            } else if (item.attributeId === Attributes.DisplayName) {
                value = UaVariant.localizedText(response.displayName);
            } else if (item.attributeId === Attributes.Description) {
                if (response.description.text.length !== 0) {
                    value = UaVariant.localizedText(response.description);
                } else {
                    statusCode = makeUaStatusCode(StatusCodes.BadNodeAttributesInvalid);
                }
            } else if (item.attributeId === Attributes.DataType) {
                if (response.dataTypeId !== null) {
                    value = UaVariant.nodeId(response.dataTypeId);
                } else {
                    statusCode = makeUaStatusCode(StatusCodes.BadNodeAttributesInvalid);
                }
            } else if (item.attributeId === Attributes.ValueRank) {
                if (response.valueRank !== null) {
                    value = UaVariant.integer(response.valueRank, UaVariantType.Int32);
                } else {
                    statusCode = makeUaStatusCode(StatusCodes.BadNodeAttributesInvalid);
                }
            } else if (item.attributeId === Attributes.AccessLevel || item.attributeId === Attributes.UserAccessLevel) {
                if (response.accessLevel !== null) {
                    value = UaVariant.integer(response.accessLevel, UaVariantType.Byte);
                } else {
                    statusCode = makeUaStatusCode(StatusCodes.BadNodeAttributesInvalid);
                }
            } else if (item.attributeId === Attributes.Historizing) {
                if (response.historizing !== null) {
                    value = UaVariant.boolean(response.historizing);
                } else {
                    statusCode = makeUaStatusCode(StatusCodes.BadNodeAttributesInvalid);
                }
            } else if (item.attributeId === Attributes.Value) {
                if (response.value !== null) {
                    value = response.value;
                } else {
                    statusCode = makeUaStatusCode(StatusCodes.BadNodeAttributesInvalid);
                }
            } else if (item.attributeId === Attributes.NodeId) {
                value = UaVariant.nodeId(item.nodeId);
            } else if (item.attributeId === Attributes.BrowseName) {
                value = UaVariant.qualifiedName(response.browseName);
            } else if (item.attributeId === Attributes.Executable || item.attributeId === Attributes.UserExecutable) {
                value = UaVariant.boolean(true);
            } else if (item.attributeId === Attributes.WriteMask || item.attributeId === Attributes.UserWriteMask) {
                value = UaVariant.integer(0, UaVariantType.UInt32);
            } else {
                statusCode = makeUaStatusCode(StatusCodes.BadAttributeIdInvalid);
            }

            this._results.push(new UaDataValue(value, statusCode));
        }
    }
}